op-e2e/sequencer_failover_test.go

```diff
@@ -17,3 +21,125 @@ func TestSequencerFailover_SetupCluster(t *testing.T) {
 		require.NotNil(t, con, "Expected conductor to be non-nil")
 	}
 }
+
+// [Category: conductor rpc]
+// In this test, we test all rpcs exposed by conductor.
+func TestSequencerFailover_ConductorRPC(t *testing.T) {
+	ctx := context.Background()
+	sys, conductors := setupSequencerFailoverTest(t)
+	defer sys.Close()
+
+	// SequencerHealthy, Leader, AddServerAsVoter are used in setup already.
+
+	// Test ClusterMembership
+	t.Log("Testing ClusterMembership")
+	c1 := conductors[Sequencer1Name]
+	c2 := conductors[Sequencer2Name]
+	c3 := conductors[Sequencer3Name]
+	membership, err := c1.client.ClusterMembership(ctx)
+	require.NoError(t, err)
+	require.Equal(t, 3, len(membership), "Expected 3 members in cluster")
+	ids := make([]string, 0)
+	for _, member := range membership {
+		ids = append(ids, member.ID)
+		require.Equal(t, consensus.Voter, member.Suffrage, "Expected all members to be voters")
+	}
+	sort.Strings(ids)
+	require.Equal(t, []string{Sequencer1Name, Sequencer2Name, Sequencer3Name}, ids, "Expected all sequencers to be in cluster")
+
+	// Test Active & Pause & Resume
+	t.Log("Testing Active & Pause & Resume")
+	active, err := c1.client.Active(ctx)
+	require.NoError(t, err)
+	require.True(t, active, "Expected conductor to be active")
+
+	err = c1.client.Pause(ctx)
+	require.NoError(t, err)
+	active, err = c1.client.Active(ctx)
+	require.NoError(t, err)
+	require.False(t, active, "Expected conductor to be paused")
+
+	err = c1.client.Resume(ctx)
+	require.NoError(t, err)
+	active, err = c1.client.Active(ctx)
+	require.NoError(t, err)
+	require.True(t, active, "Expected conductor to be active")
+
+	// Test LeaderWithID
+	t.Log("Testing LeaderWithID")
+	leader1, err := c1.client.LeaderWithID(ctx)
+	require.NoError(t, err)
+	leader2, err := c2.client.LeaderWithID(ctx)
+	require.NoError(t, err)
+	leader3, err := c3.client.LeaderWithID(ctx)
+	require.NoError(t, err)
+	require.Equal(t, leader1.ID, leader2.ID, "Expected leader ID to be the same")
+	require.Equal(t, leader1.ID, leader3.ID, "Expected leader ID to be the same")
+
+	// Test TransferLeader & TransferLeaderToServer
+	t.Log("Testing TransferLeader")
+	lid, leader := findLeader(t, conductors)
+	err = leader.client.TransferLeader(ctx)
+	require.NoError(t, err, "Expected leader to transfer leadership")
+	require.NoError(t, waitForLeadershipChange(t, leader, false))
+	ensureOnlyOneLeader(t, sys, conductors)
+	newLeader, err := leader.client.LeaderWithID(ctx)
+	require.NoError(t, err)
+	require.NotEmpty(t, newLeader.ID)
+	require.NotEqual(t, lid, newLeader.ID, "Expected leader to change")
+	require.NoError(t, waitForSequencerStatusChange(t, sys.RollupClient(newLeader.ID), true))
```

Comment:
previous revert is caused by a race condition in leader election. Turns out `LeaderWithID` can return empty string if leader is not elected yet.

And in this case, if `newLeader.ID` is empty, then `sys.RollupClient(newLeader.ID)` will be nil and cause the underlying function to panic.

We solve this by adding `ensureOnlyOneLeader(t, sys, conductors)` before to ensure that a leader is elected

Reply:
I think we should move this logic into a helper function rather than doing it in the test. Otherwise the next test we write is likely to make the same mistake of trying to use a returned leader ID without the waiter.  Whereas if we make the test code higher level, we can have a method like`WaitForLeaderActivated` that would do all of:

```
	require.NoError(t, waitForLeadershipChange(t, leader, false))
 	ensureOnlyOneLeader(t, sys, conductors)
	newLeader, err := leader.client.LeaderWithID(ctx)
 	require.NoError(t, err)
 	require.NotEmpty(t, newLeader.ID)
 	require.NotEqual(t, lid, newLeader.ID, "Expected a new leader")
 	require.NoError(t, waitForSequencerStatusChange(t, sys.RollupClient(newLeader.ID), true))
```

Maybe all of that just goes in `waitForLeadershipChange` since any test that waits for the leadership change should also wait for the new sequencer to actually start too.

Reply:
Good idea. In my PR taking over this work, I split out `waitForLeadership` and `waitForLeadershipChange`. The latter is a reusable way to wait for leadership to change, confirm it was in fact moved and not equal to previous leader, and return the new leader ID (for use when we want to transfer leadership to a specific instance and confirm that behavior)